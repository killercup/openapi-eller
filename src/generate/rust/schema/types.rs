#[derive(Debug, Clone)]
pub enum RustType {
    PlainEnum { name: TypeIdent, variants: Vec<PlainEnumVariant>, attributes: ContainerAttributes },
    DataEnum { name: TypeIdent, variants: Vec<DataEnumVariant>, attributes: ContainerAttributes },
    Struct { name: TypeIdent, fields: Vec<StructField>, attributes: ContainerAttributes },
}

#[derive(Debug, Clone)]
pub struct PlainEnumVariant {
    pub name: TypeIdent,
    pub attributes: VariantAttributes,
}

#[derive(Debug, Clone)]
pub struct DataEnumVariant {
    pub name: TypeIdent,
    pub attributes: VariantAttributes,
    pub fields: DataEnumFields,
}

#[derive(Debug, Clone)]
pub enum DataEnumFields {
    Named { fields: Vec<StructField> },
    Unnamed { fields: Vec<TypeName> },
}

#[derive(Debug, Clone)]
pub struct StructField {
    pub name: FieldName,
    pub attributes: FieldAttributes,
    pub type_name: TypeName,
    pub optional: bool,
}

#[derive(Debug, Clone)]
pub struct TypeIdent {
    pub raw: String,
    pub ident: syn::Ident,
}

#[derive(Debug, Clone)]
pub struct FieldName {
    pub raw: String,
    pub ident: syn::Ident,
}

#[derive(Debug, Clone)]
pub struct TypeName {
    pub raw: String,
    pub ident: syn::Type,
}

#[derive(Debug, Clone)]
pub struct ContainerAttributes {
    pub rename: Option<String>,
}

#[derive(Debug, Clone)]
pub struct VariantAttributes {
    pub rename: Option<String>,
}

#[derive(Debug, Clone)]
pub struct FieldAttributes {
    pub rename: Option<String>,
}
