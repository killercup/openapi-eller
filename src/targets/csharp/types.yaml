integer:
  null: long
  int64: long
  int32: int
  unix-time: Date
string:
  null: string
  date-time: DateTime
  date: DateTime
  byte: byte[]
  binary: string
  password: string
  uri: Uri
  url: Uri
number:
  null: double
  float: float
  double: double
boolean:
  null: bool
array: "List<{value}>"
map:
  null: "Dictionary<{key}, {value}>"
set: "Set<{value}>"
null: void
object:
  null: object