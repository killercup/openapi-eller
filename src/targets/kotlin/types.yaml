integer:
  null: Long
  int64: Long
  int32: Int
  unix-time: Date
string:
  null: String
  date-time: Date
  date: Date
  byte: String
  binary: String
  password: String
  uri: Url
  url: Url
number:
  null: Double
  float: Float
  double: Double
boolean:
  null: Boolean
array: "List<{value}>"
map:
  null: "Map<{key}, {value}>"
  ordered-map: "LinkedHashMap<{key}, {value}>"
set: "Set<{value}>"
null: Unit
object:
  any: Any
  null: Any