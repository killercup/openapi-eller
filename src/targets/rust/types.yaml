integer:
  null: isize
  int64: i64
  int32: i32
  unix-time: Date
string:
  null: String
  date-time: Date
  date: Date
  byte: String
  binary: String
  password: String
  uri: Url
  url: Url
number:
  null: f64
  float: f32
  double: f64
boolean:
  null: bool
array: "Vec<{value}>"
map:
  null: "BTreeMap<{key}, {value}>"
set:
  null: "HashSet<{value}>"
null: ()
object:
  null: Any
