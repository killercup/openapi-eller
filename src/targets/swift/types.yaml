integer:
  null: Int
  int64: Int64
  int32: Int32
  uint64: UInt64
  uint32: UInt32
string:
  null: String
  date-time: Date
  date: Date
  byte: String
  binary: String
  password: String
  uri: URL
  url: URL
number:
  null: Double
  float: Float
  double: Double
boolean:
  null: Bool
array: "[{value}]"
map:
  null: "[{key}: {value}]"
set: "Set<{value}>"
null: Void
object:
  null: Any