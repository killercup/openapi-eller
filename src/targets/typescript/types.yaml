integer:
  null: number
  int64: number
  int32: number
  unix-time: number
string:
  null: string
  date-time: Date
  date: Date
  byte: string
  binary: string
  password: string
  uri: URL
  url: URL
number:
  null: number
  float: number
  double: number
boolean:
  null: Boolean
array: "{value}[]"
map: "{ [key: {key}]: {value} }"
# set: "Set<{value}>"
set: "{value}[]"
null: void
object:
  any: any
  null: any