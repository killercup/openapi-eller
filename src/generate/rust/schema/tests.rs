use super::types::*;
use proc_macro2::TokenStream;
use quicli::prelude::*;
use quote::quote;
use std::{convert::TryInto, io::Write, path::Path};
use tempfile::NamedTempFile;

#[test]
fn simple_struct() -> CliResult {
    let _ = env_logger::builder().is_test(true).filter(None, log::LevelFilter::Debug).try_init();

    let schemas = vec![RustType::Struct {
        name: "Foo".try_into()?,
        fields: vec![
            StructField {
                name: "one".try_into()?,
                attributes: FieldAttributes { rename: None },
                type_name: "String".try_into()?,
            },
            StructField {
                name: "two".try_into()?,
                attributes: FieldAttributes { rename: None },
                type_name: "std::path::PathBuf".try_into()?,
            },
        ],
        attributes: ContainerAttributes { rename: None },
    }];

    insta::assert_display_snapshot_matches!("simple_struct", generate_rust(&schemas)?);

    Ok(())
}

#[test]
fn plain_enum() -> CliResult {
    let _ = env_logger::builder().is_test(true).filter(None, log::LevelFilter::Debug).try_init();

    let schemas = vec![RustType::PlainEnum {
        name: "Foo".try_into()?,
        variants: vec![PlainEnumVariant {
            name: "Bar".try_into()?,
            attributes: VariantAttributes { rename: None },
        }],
        attributes: ContainerAttributes { rename: None },
    }];

    insta::assert_display_snapshot_matches!("plain_enum", generate_rust(&schemas)?);

    Ok(())
}

#[test]
fn data_enum() -> CliResult {
    let _ = env_logger::builder().is_test(true).filter(None, log::LevelFilter::Debug).try_init();

    let schemas = vec![RustType::DataEnum {
        name: "Foo".try_into()?,
        variants: vec![
            DataEnumVariant {
                name: "Bar".try_into()?,
                attributes: VariantAttributes { rename: None },
                fields: DataEnumFields::Named {
                    fields: vec![
                        StructField {
                            name: "one".try_into()?,
                            attributes: FieldAttributes { rename: None },
                            type_name: "String".try_into()?,
                        },
                        StructField {
                            name: "two".try_into()?,
                            attributes: FieldAttributes { rename: None },
                            type_name: "std::path::PathBuf".try_into()?,
                        },
                    ],
                },
            },
            DataEnumVariant {
                name: "Baz".try_into()?,
                attributes: VariantAttributes { rename: None },
                fields: DataEnumFields::Unnamed {
                    fields: vec!["std::borrow::Cow<'static, str>".try_into()?],
                },
            },
        ],
        attributes: ContainerAttributes { rename: None },
    }];

    insta::assert_display_snapshot_matches!("data_enum", generate_rust(&schemas)?);

    Ok(())
}

fn generate_rust(schemas: &[RustType]) -> Result<String, Error> {
    let schemas: TokenStream = schemas.iter().map(|x| quote! { #x }).collect();

    let mut target = NamedTempFile::new()?;
    writeln!(&mut target, "{}", schemas)?;
    rustfmt(&target.path())?;

    read_file(&target.path())
}

fn rustfmt(path: &Path) -> Result<(), Error> {
    let fmt = std::process::Command::new("rustfmt").arg(&path).status()?;
    if !fmt.success() {
        Err(format_err!("rustfmt failed"))?;
    }
    Ok(())
}
