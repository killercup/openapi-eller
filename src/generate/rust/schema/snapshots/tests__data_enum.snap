---
created: "2019-06-03T21:30:16.807337Z"
creator: insta@0.8.1
source: src/generate/rust/schema/tests.rs
expression: generate_rust(&schemas)?
---
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Foo {
    Bar {
        one: String,
        two: std::path::PathBuf,
    },
    Baz(std::borrow::Cow<'static, str>),
}

